distributed/process.py

```diff
@@ -107,70 +97,61 @@ def _run(cls, target, args, kwargs):
-            except Empty:
-                pass
+        while True:
+            msg = q.get()
+            logger.debug("[%s] got message %r" % (r, msg))
+            op = msg['op']
+            if op == 'start':
+                _call_and_set_future(loop, msg['future'], _start)
+            elif op == 'terminate':
+                _call_and_set_future(loop, msg['future'], process.terminate)
+            elif op == 'stop':
+                break
             else:
-                logger.debug("[%s] got message %r" % (r, msg))
-                op = msg['op']
-                if op == 'start':
-                    _call_and_set_future(loop, msg['future'], _start)
-                elif op == 'terminate':
-                    _call_and_set_future(loop, msg['future'], process.terminate)
-                elif op == 'stop':
-                    raise SystemExit
-                else:
-                    assert 0, msg
-
-        def _maybe_notify_exit(exitcode):
+                assert 0, msg
+
+    @classmethod
+    def _watch_process(cls, selfref, process, state, q):
+        r = repr(selfref())
+        process.join()
+        exitcode = process.exitcode
+        if exitcode is not None:
+            logger.debug("[%s] process %r exited with code %r",
+                         r, state.pid, exitcode)
+            state.is_alive = False
+            state.exitcode = exitcode
+            # Make sure the process is removed from the global list
+            # (see _children in multiprocessing/process.py)
+            process.join(timeout=0)
+            # Then notify the Process object
             self = selfref()  # only keep self alive when required
             try:
                 if self is not None:
                     self._loop.add_callback(self._on_exit, exitcode)
             finally:
                 self = None  # lose reference
-
-        while True:
-            # Periodic poll as there's no simple way to poll a threading Queue
-            # and a mp Process at the same time.
-            try:
-                _process_one_message(timeout=0.02)
-            except SystemExit:
-                return
-            # Did process end?
-            exitcode = process.exitcode
-            if exitcode is not None:
-                logger.debug("[%s] process %r exited with code %r",
-                             r, state.pid, exitcode)
-                state.is_alive = False
-                state.exitcode = exitcode
-                # Make sure the process is removed from the global list
-                # (see _children in multiprocessing/process.py)
-                process.join(timeout=0)
-                # Then notify the Process object
-                _maybe_notify_exit(exitcode)
-                break
-
-        while True:
-            # Idle poll
-            try:
-                _process_one_message(timeout=1e6)
-            except SystemExit:
-                return
+                q.put({'op': 'stop'})
```

Comment:
I don't immediately see an issue here.  All `{'op': 'stop'}` does is stop the message queue thread, so this wouldn't have a large impact.  You know this part of the code better than I do though.  If you see something amiss then please speak up.